### PR TITLE
Adding html_description attribute to the Cache object

### PR DESCRIPTION
**Pull request by DarkOutcast6**
As stated in issue #216, you cannot currently acces the raw HTML version of a description.
So, I'm working on adding a `html_description` attribute to the Cache object. I understand it may not work, but I'm semi-confident I understand Beautiful Soup well enough and added the right code. Bear with me though, as the code could definitely contain mistakes.

### CODE REVIEW

**Comment by FriedrichFroebel** [comment]
Thanks for the test. Could you please add a check for the actual parsing from the HTML as well by extending one of the existing cassette-based tests? Additionally, CI is currently failing due to code style violations.

Do you have a solution for this?

> Additionally, do you see any way to avoid doing the same tree querying twice for both the `description` and `description_html` attribute?

**Comment by DarkOutcast6** [comment]
I'm not entirely sure what you mean by the following. (I apologize, as I am a novice with these type of things)
> Could you please add a check for the actual parsing from the HTML as well by extending one of the existing cassette-based tests?

------------------------------

I think I understand what you mean by 
> avoid doing the same tree querying twice for both the description and description_html attribute?

If I _do_ understand you correctly, I'm working on it now.

**Comment by DarkOutcast6** [comment]
I'm not entirely sure what you mean by the following. (I apologize, as I am a novice with these type of things)
> Could you please add a check for the actual parsing from the HTML as well by extending one of the existing cassette-based tests?

------------------------------

I think I undertsnad what you mean by 
> avoid doing the same tree querying twice for both the description and description_html attribute?

If I _do_ understand you correctly, I'm working on it now.

**Comment by FriedrichFroebel** [comment]
> I'm not entirely sure what you mean by the following. (I apologize, as I am a novice with these type of things)
 >>   Could you please add a check for the actual parsing from the HTML as well by extending one of the existing cassette-based tests?

Your current test only covers explicitly setting the `description_html` field in one case, but does not ensure that the parsed data from the Geocaching.com page is indeed correct.  AFAIK this is not tested for the `description` field either, but checking this makes sense.

> I think I understand what you mean by
>> avoid doing the same tree querying twice for both the description and description_html attribute?
> If I do understand you correctly, I'm working on it now.

You only fixed this for exactly one case, but not for the remaining ones where the same approach might make sense as well.

Additionally, when you look at the CI runs here, you will see that there currently are code style violations. And https://github.com/tomasbedrich/pycaching/pull/218/files#diff-c1861c7962a2fda2a943282daf8c8cb0030863a5cfdf8acc43402e3f1b381857R157 is considered both a name violation according to PEP8 and invalid variable name syntax.

**Comment by DarkOutcast6** [comment]
Sorry for the trouble -- I feel so close :)

Pretty sure I figured out what the code violation is. (correct me if I didn't fix it or if there is more)

___

> You only fixed this for exactly one case, but not for the remaining ones where the same approach might make sense as well.

I looked over the code and can't find another case where this occurs.

___

> Your current test only covers explicitly setting the description_html field in one case, but does not ensure that the parsed data from the Geocaching.com page is indeed correct. AFAIK this is not tested for the description field either, but checking this makes sense.

I understand what you mean, but how would I actually do this?

___

Sorry for sounding being an idiot through this whole procces lol. Hope I'm not too annoying! :wink:

**Comment by FriedrichFroebel** [comment]
> Pretty sure I figured out what the code violation is. (correct me if I didn't fix it or if there is more)

Your variable names should use an underscore, not a dash (this is invalid Python syntax). Additionally, CI is reporting an empty line too much. This needs to be fixed.

> I looked over the code and can't find another case where this occurs.

You added changes to https://github.com/tomasbedrich/pycaching/pull/218/files#diff-c1861c7962a2fda2a943282daf8c8cb0030863a5cfdf8acc43402e3f1b381857R832 and https://github.com/tomasbedrich/pycaching/pull/218/files#diff-c1861c7962a2fda2a943282daf8c8cb0030863a5cfdf8acc43402e3f1b381857R958 as well, not only at https://github.com/tomasbedrich/pycaching/pull/218/files#diff-c1861c7962a2fda2a943282daf8c8cb0030863a5cfdf8acc43402e3f1b381857L157 You should use the same approach all over.

> I understand what you mean, but how would I actually do this?

There are multiple tests in `test.test_cache.TestMethods`. Consider extending the tests for the three loader methods to test for `description` (if currently missing) and `description_html`, ensuring that the HTML variant includes HTML tags, while the plain-text one does not. https://github.com/tomasbedrich/pycaching/blob/bef4ca07693d4d1e06b070f4c72a154ca3f487a4/test/test_cache.py#L282 is an example of an existing simple `description` test where the existing assertion probably should be expanded to cover a case where HTML tags have been removed (which is not the case here yet).

**Comment by DarkOutcast6** [comment]
What is the difference between ` self.assertIn()`, ` self.assertEqual()` and ` self.assertGreater()` and which of these should be used for description_html?

**Comment by FriedrichFroebel** [comment]
I recommend you to familiarize yourself with general `unittest`-based testing (which is the stdlib Python module for testing) to do so. You will find the official documentation for the different assertions at https://docs.python.org/3/library/unittest.html#assert-methods

**Comment by DarkOutcast6** [comment]
What else do I need to add for this to be valid?
It be very helpful if you could give me a list of things I need to edit.

**Comment by FriedrichFroebel** [comment]
* The code style needs to be fixed. See the CI runs for this or run `black` in your local development environment to check/fix this.
* As mentioned in the second section at https://github.com/tomasbedrich/pycaching/pull/218#issuecomment-1712584664, please consider re-using the retrieved tree for the text and string rendering as you have done for the other case.
* As mentioned in the third section at https://github.com/tomasbedrich/pycaching/pull/218#issuecomment-1712584664, please extend the tests for the three loader methods you changed with suitable tests (the test you added in your latest commit would probably fail for this specific geocache instance). To do so, please have a look at the original descriptions of the corresponding geocaches used for testing purposes.
  * Add a corresponding assertion for a part of the `description_html` property, ensuring that it does indeed include the correct HTML tags.
  * Improve/rewrite the current assertion for the `description` property to check for exactly the same section as in `description_html` with the only difference being that the expected content does indeed not contain the corresponding HTML tags.

**Comment by DarkOutcast6** [comment]
>The code style needs to be fixed.

I fixed the `black` code styling issues.

___

>  re-using the retrieved tree for the text and string rendering as you have done for the other case.

Made it so that it now re-uses the tree.

___

The CI runs now state that there is a problem with `flake8`. The error reads: 

./pycaching/cache.py:628:5: F8[11](https://github.com/tomasbedrich/pycaching/actions/runs/6166084369/job/16740063682?pr=218#step:7:12) redefinition of unused 'description_html' from line 619
    @description.setter
    ^
1     F811 redefinition of unused 'description_html' from line 619

I don't fully understand what it means. Could you explain what it means?

File: pycaching/cache.py [hunk 1]
```diff
@@ -1020,7 +1022,7 @@ def _get_log_counts_from_print_page(soup):
         if not element:
             raise errors.ValueError("Log counts could not be found.")
 
-        # Text gives numbers and verbose descriptions of the current values as well as an
+        # Text gives numbers and verbose s of the current values as well as an
```
Comment (FriedrichFroebel) [inline pycaching/cache.py:1025]: There is a typo.

File: pycaching/cache.py [hunk 2]
```diff
@@ -613,6 +616,20 @@ def description(self, description):
...
+    @property
+    @lazy_loaded
+    def description_html(self):
+        """The cache long description in raw HTML.
+
+        :type: :class:`str`
+        """
+        return self._description
+
+    @description.setter
```
Comment (FriedrichFroebel) [inline pycaching/cache.py:628]: This should be the setter for the `description_html` property. At the moment, you have two setters for `description`.

File: pycaching/cache.py (hunk 1 above)
Comment (DarkOutcast6) [reply pycaching/cache.py:1025]: I don’t understand — I did not alter line 1020 - 1025.

File: pycaching/cache.py (hunk 2 above)
Comment (DarkOutcast6) [reply pycaching/cache.py]: Just fixed!

File: pycaching/cache.py (hunk 1 above)
Comment (FriedrichFroebel) [reply pycaching/cache.py]: You can see from the referenced diff that `descriptions` has been there before which still should be kept.

File: pycaching/cache.py [hunk 3]
```diff
@@ -153,8 +153,10 @@ def _from_print_page(cls, geocaching, guid, soup):
         cache_info["attributes"] = {attr_name: attr_setting == "yes" for attr_name, _, attr_setting in attributes}
         if "attribute" in cache_info["attributes"]:  # 'blank' attribute
             del cache_info["attributes"]["attribute"]
-        cache_info["summary"] = content.find("h2", string="Short Description").find_next("div").text
-        cache_info["description"] = content.find("h2", string="Long Description").find_next("div").text
+        cache_info["summary"] = content.find("h2", string="Short ").find_next("div").text
```
Comment (FriedrichFroebel) [inline pycaching/cache.py:156]: Why did you change this string from "Short Description" to "Short "? This seems to be wrong and lets the existing tests fail.

**Comment by FriedrichFroebel** [comment]
Thanks for the fixes. I have added two other comments to the diff. Additionally, the required tests are still missing or incomplete and should be implemented as requested previously.

File: pycaching/cache.py (hunk 3 above)
Comment (DarkOutcast6) [reply pycaching/cache.py:156]: I didn’t. That’s so weird :/

File: pycaching/cache.py (hunk 1 above)
Comment (DarkOutcast6) [reply pycaching/cache.py]: Fixed it!

File: pycaching/cache.py [hunk 4]
```diff
@@ -613,6 +616,20 @@ def description(self, description):
...
         self._description = description
 
+    @property
+    @lazy_loaded
+    def description_html(self):
+        """The cache long description in raw HTML.
+
+        :type: :class:`str`
+        """
+        return self._description
```
Comment (FriedrichFroebel) [inline pycaching/cache.py:626]: I just have  been confused because of the CI results, but this should be `return self._description_html` of course.

Comment (DarkOutcast6) [reply pycaching/cache.py]: Just fixed! Thanks for pointing that typo out!

**Comment by DarkOutcast6** [comment]
> Additionally, the required tests are still missing or incomplete and should be implemented as requested previously.

Working on this now, hopefully will be finished this week!
If there is anything else that needs to be brought to my attention, feel free to tell me now!

**Comment by DarkOutcast6** [comment]
> Add a corresponding assertion for a part of the description_html property, ensuring that it does indeed include the correct HTML tags.
Just fixed it!

Hypothetically, this should work now, correct? If so, is there anything else I need to update? (And if it isn't, I'll update accordingly)

**Comment by FriedrichFroebel** [comment]
You implemented the new property for the three different loading methods. I can find (at most) only two tests which check that loading the data does indeed work correctly for `description_html`. We should cover all three cases.

I previously requested that the corresponding tests for the `description` should be adopted to actually show that they do not contain HTML tags. This is not the case at the moment, thus we cannot see from the tests if `description` does indeed not contain any HTML tags, as there is nothing to strip for the existing text. Ideally, your test snippet for `description` and `description_html` would be identical with only the tags differing, while having HTML tags in-between as well, not only at the borders.

Additionally, linting currently fails due to the long line. Do you really need such a long text for the test or could you shorten this to still cover the relevant cases (see previous paragraph), but not being too verbose?

**Comment by DarkOutcast6** [comment]
> description should be adopted to actually show that they do not contain HTML tags. This is not the case at the moment, thus we cannot see from the tests if description does indeed not contain any HTML tags, as there is nothing to strip for the existing text. Ideally, your test snippet for description and description_html would be identical with only the tags differing, while having HTML tags in-between as well, not only at the borders.
Will work on this now.

_____

> Additionally, linting currently fails due to the long line.
Fixing now.

____

> You implemented the new property for the three different loading methods. I can find (at most) only two tests which check that loading the data does indeed work correctly for description_html. We should cover all three cases.
Not really sure what you mean...?

**Comment by FriedrichFroebel** [comment]
You changed `_from_print_page`, `load_quick` and `load_by_guid`. You might add a new test for `_from_print_page` (which does not seem to exist at the moment), while extending `test.test_cache.TestMethods.test_load` and `test.test_cache.TestMethods.test_load_by_guid`.

File: test/test_cache.py [hunk 5]
```diff
@@ -33,7 +33,8 @@ def setUp(self):
             hidden=date(2000, 1, 1),
             attributes={"onehour": True, "kids": False, "available": True},
             summary="text",
-            description="long text",
+            description="engelmz",
+            description_html="<b><font color="red">engelmz</font></b>",
```
Comment (FriedrichFroebel) [inline test/test_cache.py:37]: This is invalid syntax and wrongly repeated twice below. AFAIK the `black` formatter prefers `'<b><font color="red">engelmz</font></b>'` in these cases.

Comment (DarkOutcast6) [reply test/test_cache.py:37]: Pretty sure I just fixed it.

**Comment by DarkOutcast6** [comment]
Looks like it's working! 

___

By the way, what does 
image
mean and is it important? 

___

> You changed _from_print_page, load_quick and load_by_guid. You might add a new test for _from_print_page (which does not seem to exist at the moment), while extending test.test_cache.TestMethods.test_load and test.test_cache.TestMethods.test_load_by_guid. 

Do I still need to this? If so, what specifically must I change/add?

**Review by FriedrichFroebel — APPROVED** [review]
Thanks for your patience. I have added some small enhancements myself to improve the tests itself.

**Comment by DarkOutcast6** [comment]
Thank you so much for the help!
When will this be released?

**Comment by DarkOutcast6** [comment]
Thank you so much for the help, @FriedrichFroebel! Do you know when this will be added to a release?

**Comment by FriedrichFroebel** [comment]
Version 4.4.0 has been released including this new feature.
